core/src/main/java/org/kiwix/kiwixmobile/core/webserver/ZimHostActivity.kt

```diff
@@ -0,0 +1,290 @@
+          Log.v(tag, "ZIM PATH : " + file.absolutePath)
+        }
+      }
+      return selectedBooksPath
+    }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_zim_host)
+
+    setUpToolbar()
+
+    bookDelegate = BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil, null, null,
+      { bookOnDiskItem ->
+        select(bookOnDiskItem)
+        Unit
+      })
+    bookDelegate.selectionMode = SelectionMode.MULTI
+    booksAdapter = BooksOnDiskAdapter(
+      bookDelegate,
+      BookOnDiskDelegate.LanguageDelegate
+    )
+    if (savedInstanceState != null) {
+      ip = savedInstanceState.getString(ipStateKey)
+      layoutServerStarted()
+    }
+    recyclerViewZimHost.adapter = booksAdapter
+    presenter.attachView(this)
+
+    serviceConnection = object : ServiceConnection {
+
+      override fun onServiceConnected(className: ComponentName, service: IBinder) {
+        hotspotService = (service as HotspotService.HotspotBinder).service
+        hotspotService!!.registerCallBack(this@ZimHostActivity)
+      }
+
+      override fun onServiceDisconnected(arg0: ComponentName) {}
+    }
+
+    startServerButton.setOnClickListener { startStopServer() }
+  }
+
+  private fun startStopServer() {
+    when {
+      ServerUtils.isServerStarted -> stopServer()
+      selectedBooksPath.size > 0 -> startHotspotManuallyDialog()
+      else -> Toast.makeText(
+        this,
+        R.string.no_books_selected_toast_message,
+        Toast.LENGTH_SHORT
+      ).show()
+    }
+  }
+
+  private fun stopServer() {
+    startService(createHotspotIntent(ACTION_STOP_SERVER))
+  }
+
+  private fun select(bookOnDisk: BooksOnDiskListItem.BookOnDisk) {
+    val booksList: ArrayList<BooksOnDiskListItem> = ArrayList()
+    for (item in booksAdapter.items) {
+      if (item == bookOnDisk) {
+        item.isSelected = !item.isSelected
+      }
+      booksList.add(item)
+    }
+    booksAdapter.items = booksList
+    saveHostedBooks(booksList)
+  }
+
+  override fun onStart() {
+    super.onStart()
+    bindService()
+  }
+
+  override fun onStop() {
+    super.onStop()
+    unbindService()
+  }
+
+  private fun bindService() {
+    bindService(
+      Intent(this, HotspotService::class.java), serviceConnection,
+      Context.BIND_AUTO_CREATE
+    )
+  }
+
+  private fun unbindService() {
+    if (hotspotService != null) {
+      unbindService(serviceConnection)
+      hotspotService!!.registerCallBack(null)
+    }
+  }
+
+  override fun onResume() {
+    super.onResume()
+    presenter.loadBooks(sharedPreferenceUtil.hostedBooks)
+    if (ServerUtils.isServerStarted) {
+      ip = ServerUtils.getSocketAddress()
+      layoutServerStarted()
+    }
+  }
+
+  private fun saveHostedBooks(booksList: ArrayList<BooksOnDiskListItem>) {
+    sharedPreferenceUtil.hostedBooks =
+      booksList.filter { book -> book.isSelected && book is BookOnDisk && book.book.title != null }
```

Comment:
use `filterIsInstance` and `mapNotNull` to prettify this a bit

Reply:
https://github.com/kiwix/kiwix-android/pull/1569/commits/0c084510ea8f25ef9801a9615cdf0f15c43c53e0